File: source/utility.c
// Stratus: utility.c
// (c) 2026 Connor J. Link. All Rights Reserved.

#include <stdint.h>
#include <stdarg.h>
#include <stdbool.h>

#include "utility.h"

static inline void platform_putchar(char c)
{
    /* UART0 on QEMU virt */
    volatile unsigned char* thr = (volatile unsigned char*)0x10000000u;
    volatile unsigned char* lsr = (volatile unsigned char*)0x10000005u;

    while (((*lsr) & (1u << 5)) == 0) { }
    *thr = (unsigned char)c;
}

size_t max(size_t x, size_t y)
{
    return x > y ? x : y;
}

size_t min(size_t x, size_t y)
{
    return y > x ? x : y;
}

size_t strlen(const char* str)
{
    size_t len = 0;

    while (str[len])
    {
        len++;
    }

    return len;
}

char* strcpy(char* destination, const char* string)
{
    char* result = destination;
    
    while (*string)
    {
        *destination++ = *string++;
    }

    *destination = '\0';
    return result;
}

int strcmp(const char* string1, const char* string2)
{
    while (*string1 && *string2)
    {
        if (*string1 != *string2)
        {
            break;
        }

        string1++;
        string2++;
    }

    return *(uint8_t*)string1 - *(uint8_t*)string2;
}

void* memset(void* pointer, unsigned char value, size_t number)
{
    uint8_t* p = (unsigned char*)pointer;

    while (number--)
    {
        *p++ = value;
    }

    return pointer;
}

void* memcpy(void* destination, const void* source, size_t number)
{
    uint8_t* d = (uint8_t*)destination;
    const uint8_t* s = (const uint8_t*)source;

    while (number--)
    {
        *d++ = *s++;
    }

    return destination;
}

void putchar(char c)
{
    if (c == '\n')
    {
        platform_putchar('\r');
    }

    platform_putchar(c);
}

void printf(const char* format, ...)
{
    va_list va;
    va_start(va, format);

    while (*format)
    {
        if (*format == '%')
        {
            format++;

            switch (*format)
            {
                case 'c':
                {
                    char c = (char)va_arg(va, int);
                    putchar(c);
                    break;
                }
                case 's':
                {
                    const char* str = va_arg(va, const char*);
                    while (*str)
                    {
                        putchar(*str++);
                    }
                    break;
                }
                case 'd':
                {
                    int number = va_arg(va, int);
                    char buffer[12];
                    int i = 0;
                    int is_negative = 0;

                    if (number < 0)
                    {
                        is_negative = 1;
                        number = -number;
                    }

                    do
                    {
                        buffer[i++] = (number % 10) + '0';
                        number /= 10;
                    } while (number > 0);

                    if (is_negative)
                    {
                        buffer[i++] = '-';
                    }

                    while (i--)
                    {
                        putchar(buffer[i]);
                    }

                    break;
                }
                case 'x':
                {
                    const unsigned int number = va_arg(va, unsigned int);
                    for (int i = (sizeof(unsigned int) * 2) - 1; i >= 0; i--)
                    {
                        unsigned char nibble = (number >> (i * 4)) & 0xF;
                        putchar("0123456789ABCDEF"[nibble]);
                    }
                }
                case '%':
                {
                    putchar('%');
                    break;
                }
                case '\0':
                {
                    putchar('%');
                    goto done;
                }
                default:
                    goto done;
            }
        }
        else
        {
            putchar(*format);
        }

        format++;
    }

    done:
        va_end(va);
}

File: src/utility.zig
// Stratus: utility.zig
// (c) 2026 Connor J. Link. All Rights Reserved.

const std = @import("std");

pub fn max(x: usize, y: usize) usize {
    return if (x > y) x else y;
}

pub fn min(x: usize, y: usize) usize {
    return if (y > x) x else y;
}

pub fn strlen(string: [*:0]const u8) usize {
    var length: usize = 0;
    while (string[length] != 0) {
        length += 1;
    }
    return length;
}

pub fn strcpy(destination: [*]u8, string: [*:0]const u8) [*]u8 {
    var local_destination = destination;
    var local_source = string;
    while (local_source[0] != 0) {
        local_destination[0] = local_source[0];
        local_destination += 1;
        local_source += 1;
    }
    local_destination[0] = 0;
    return destination;
}

pub fn strcmp(string1: [*:0]const u8, string2: [*:0]const u8) i32 {
    var local_string1 = string1;
    var local_string2 = string2;
    while (local_string1[0] != 0 and local_string2[0] != 0) {
        if (local_string1[0] != local_string2[0]) {
            break;
        }
        local_string1 += 1;
        local_string2 += 1;
    }
    return @as(i32, local_string1[0]) - @as(i32, local_string2[0]);
}

pub fn memset(pointer: [*]u8, value: u8, number: usize) [*]u8 {
    var local_pointer = pointer;
    var local_number = number;
    while (local_number > 0) {
        local_pointer[0] = value;
        local_pointer += 1;
        local_number -= 1;
    }
    return pointer;
}

pub fn memcpy(destination: [*]u8, source: [*]const u8, number: usize) [*]u8 {
    var local_destination = destination;
    var local_source = source;
    var local_number = number;
    while (local_number > 0) {
        local_destination[0] = local_source[0];
        local_destination += 1;
        local_source += 1;
        local_number -= 1;
    }
    return destination;
}

inline fn platform_putchar(c: u8) void {
    const thr = @as(*volatile u8, @ptrFromInt(0x10000000));
    const lsr = @as(*volatile u8, @ptrFromInt(0x10000005));

    while ((lsr.* & (1 << 5)) == 0) {}
    thr.* = c;
}

pub fn putchar(c: u8) void {
    if (c == '\n') {
        platform_putchar('\r');
    }
    platform_putchar(c);
}

pub fn printf(format: [*:0]const u8, args: anytype) void {
    var fmt = format;
    var arg_index: usize = 0;

    while (fmt[0] != 0) {
        if (fmt[0] == '%') {
            fmt += 1;

            switch (fmt[0]) {
                'c' => {
                    const c = args[arg_index];
                    arg_index += 1;
                    putchar(@as(u8, @intCast(c)));
                },
                's' => {
                    const str = args[arg_index];
                    arg_index += 1;
                    var s = str;
                    while (s[0] != 0) {
                        putchar(s[0]);
                        s += 1;
                    }
                },
                'd' => {
                    var number = args[arg_index];
                    arg_index += 1;
                    var is_negative = false;

                    if (number < 0) {
                        is_negative = true;
                        number = -number;
                    }

                    var buffer: [12]u8 = undefined;
                    var i: usize = 0;

                    while (i < buffer.len) {
                        buffer[i] = @as(u8, @intCast((number % 10) + '0'));
                        i += 1;
                        number /= 10;
                        if (number == 0) break;
                    }

                    if (is_negative) {
                        buffer[i] = '-';
                        i += 1;
                    }

                    while (i > 0) {
                        i -= 1;
                        putchar(buffer[i]);
                    }
                },
                'u' => {
                    var number = args[arg_index];
                    arg_index += 1;
                    var buffer: [12]u8 = undefined;
                    var i: usize = 0;

                    while (i < buffer.len) {
                        buffer[i] = @as(u8, @intCast((number % 10) + '0'));
                        i += 1;
                        number /= 10;
                        if (number == 0) break;
                    }

                    while (i > 0) {
                        i -= 1;
                        putchar(buffer[i]);
                    }
                },
                'x' => {
                    const number = args[arg_index];
                    arg_index += 1;
                    const hex_digits = "0123456789ABCDEF";
                    var shift: u32 = (@sizeOf(@TypeOf(number)) * 8) - 4;

                    while (true) {
                        const nibble = @as(u4, @intCast((number >> shift) & 0xF));
                        putchar(hex_digits[nibble]);
                        if (shift == 0) break;
                        shift -= 4;
                    }
                },
                '%' => {
                    putchar('%');
                },
                0 => {
                    putchar('%');
                    return;
                },
                else => return,
            }
        } else {
            putchar(fmt[0]);
        }

        fmt += 1;
    }
}
